src/cns.ts

```diff
@@ -0,0 +1,219 @@
+import { ResolutionErrorCode } from './resolutionError';
+import Contract from './ens/contract/contract';
+import { formatsByCoinType } from '@ensdomains/address-encoder';
+
+/**
+ * Class to support connection with Crypto naming service
+ * @param network - network string such as
+ * - mainnet
+ * - ropsten
+ * @param url - main api url such as
+ * - https://mainnet.infura.io
+ * @param registryAddress - address for a registry contract
+ */
+export default class Cns extends EtheriumNamingService {
+  readonly network: string;
+  readonly url: string;
+  readonly registryAddress?: string;
+  /** @ignore */
+  private cnsContract: any;
+  /** @ignore */
+  private web3: any;
+  /** @ignore */
+  readonly RegistryMap: RegistryMap = {
+    mainnet: '0x234Dad2a1bCfB3b66c5c6040c17112A81565493e',
+  };
+
+ /**
+   * Source object describing the network naming service operates on
+   * @param source - if specified as a string will be used as main url, if omited then defaults are used
+   * @throws ConfigurationError - when either network or url is setup incorrectly
+   */
+  constructor(source: NamingServiceSource = true) {
+    super();
+    source = this.normalizeSource(source);
+    this.network = source.network as string;
+    this.url = source.url;
+    if (!this.network) {
+      throw new Error('Unspecified network in Namicorn CNS configuration');
+    }
+    if (!this.url) {
+      throw new Error('Unspecified url in Namicorn CNS configuration');
+    }
+    this.registryAddress = source.registry
+      ? source.registry
+      : this.RegistryMap[this.network];
+    if (this.registryAddress) {
+      this.cnsContract = new Contract(
+        this.url,
+        cnsInterface,
+        this.registryAddress,
+      );
+    }
+  }
+
+  /**
+   * Checks if the domain is in valid format
+   * @param domain - domain name to be checked
+   * @returns
+   */
+  isSupportedDomain(domain: string): boolean {
+    return domain.indexOf('.') > 0 && /^.{1,}\.(crypto)$/.test(domain);
+  }
+
+  /**
+   * Resolves the given domain. [DEPRICATED]
+   * @async
+   * @param domain - domain name to be resolved
+   * @returns- Returns a promise that resolves in an object
+   */
+  async resolve(domain: string): Promise<NamicornResolution> {
+    try {    
+      this.ensureSupportedDomain(domain);
+      var [nodeHash, owner, ttl, resolver] = await this.getResolutionMeta(domain);
+    } catch(err) {
+      if (err instanceof ResolutionError)
+        return null;
+      throw err;
+    }
+    const address = await this.fetchAddress(resolver, nodeHash, 'ETH');
+    return {
+      addresses: {
+        'ETH': address,
+      },
+      meta: {
+          owner,
+          type: 'cns',
+          ttl
+      }};
+  }
+
+  /**
+   * Produces ENS namehash
+   * @param domain - domain to be hashed
+   * @return ENS namehash of a domain
+   */
+  namehash(domain: string): string {
+    this.ensureSupportedDomain(domain)
+    return hash(domain);
```

Comment:
I am seeing this method being shared between `ens` and `cns`. Can we move it to `EthereumNamingService` ?